ngsderive/commands/instrument.py

```diff
@@ -119,13 +119,16 @@ def predict_instrument_from_fcids(fcids):
 
 def resolve_instrument(possible_instruments_by_iid,
                        possible_instruments_by_fcid,
-                       at_least_one_instrument_detected):
+                       at_least_one_instrument_detected,
+                       malformed_read_names_detected):
     if len(possible_instruments_by_iid) == 0 and len(
             possible_instruments_by_fcid) == 0:
         if at_least_one_instrument_detected:
```

Comment:
@claymcleod this doesn't impact my changes, but I'm just curious what's going on here. If both `possible_instruments` sets are empty, logically shouldn't `at_least_one_instrument_detected` always be `False`?

I think I'm just not understanding what `at_least_one_instrument_detected` is meant to signify, because the logic in `predict_instrument_from_fcids()` and `predict_instrument_from_iids()` looks to me like it will always be set `True`

Reply:
If you'll notice, both `possible_instruments` are set as follows:

1. Go across the number of reads to examine (let's say 5 here) and calculate _union_ of all possible instruments based on the ID. This gives you all instruments that could have generated the read.
2. Do an _intersection_ across all of the steps generated by point 1 to get the final set of instruments that could have generated the set ("which of these instruments could have generated all of the reads here?").

The size of the set from point 2 could be zero for two different reasons:

- We could not detect any machines that could have generated even a single read.
- The intersection across the reads was a totally non-overlapping set, which would mean that reads had conflicting evidence about which machine generated them.

`at_least_one_instrument_detected` essentially checks for the second bullet here and flags the case as "multiple instruments" if it runs across it.